pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go

```diff
@@ -0,0 +1,333 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assemble
+
+import (
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	ctrlutil "github.com/oam-dev/kubevela/pkg/controller/utils"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+// NewAssembleOptions create a Options
+func NewAssembleOptions(appRevision *v1beta1.ApplicationRevision) *Options {
+	return &Options{AppRevision: appRevision}
+}
+
+// Options contains configuration to assemble resources recorded in the ApplicationRevision.
+// 'Assemble' means completing workload/trait resources and get them ready to emit into K8s
+// cluster.
+type Options struct {
+	AppRevision     *v1beta1.ApplicationRevision
+	WorkloadOptions []WorkloadOption
+
+	AppConfig      *v1alpha2.ApplicationConfiguration
+	Comps          []*v1alpha2.Component
+	AppName        string
+	AppNamespace   string
+	AppLables      map[string]string
+	AppAnnotations map[string]string
+	AppOwnerRef    *metav1.OwnerReference
+
+	assembledWorkloads map[string]*unstructured.Unstructured
+	assembledTraits    map[string][]*unstructured.Unstructured
+	referencedScopes   map[runtimev1alpha1.TypedReference][]runtimev1alpha1.TypedReference
+
+	assembled bool
+}
+
+// WorkloadOption will be applied to each workloads AFTER it has been assembled by generic rules shown below:
+// 1) use component name as workload name
+// 2) use application namespace as workload namespace if unspecified
+// 3) set application as workload's owner
+// 4) pass all application's lables and annotations to workload's
```

Comment:
```suggestion
// 4) pass all application's labels and annotations to workload's
```

Please clean all the mis-spell